Issue #30 : Add S3 inbound streaming channel adapter.

tzolov opened the pull request:
This implementation is meant to resolve #30 

* Add S3InboundStreamingMessageSource and related S3FileInfo
* Extend spring-integration-aws-1.1.xsd with s3-inbound-streaming-channel-adapter tag.
* Add S3StreamingInboundChannelAdapterParser and s3-inbound-streaming-channel-adapter handler.
* Add S3 streaming spring tests.
* Add S3 streaming documentation to README.md.

artembilan commented:
Excellent, @tzolov !

Only two minor concerns before proper review:
1. This PR fully isn't about that JIRA. It is exactly for the #30 and just for the S3. Nothing here about (S)FTP
2. We have to modify XSD for `1.1` version already. Sorry, that is my fault. Will address soon :smile: After that you have to rebase and apply XSD changes properly to 1.1

artembilan commented:
XSD has been fixed: https://github.com/spring-projects/spring-integration-aws/commit/de4dfed908c186129bda530000c5607a7d99106d

artembilan commented on src/main/java/org/springframework/integration/aws/config/xml/S3InboundChannelAdapterParser.java:
```diff
@@ -50,5 +50,4 @@ protected String getMessageSourceClassname() {
 	protected Class<? extends InboundFileSynchronizer> getInboundFileSynchronizerClass() {
 		return S3InboundFileSynchronizer.class;
 	}
-
```
No. Blank like before the last curly brace in the class: https://github.com/spring-projects/spring-integration/wiki/Spring-Integration-Framework-Code-Style.
But, eh, that's minor. Will fix on merge :smile:

artembilan commented on src/main/java/org/springframework/integration/aws/config/xml/S3InboundStreamingChannelAdapterParser.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Parser for the AWS 's3-inbound-streaming-channel-adapter' element.
+ *
+ * @author Christian Tzolov
```
`@since 1.1`

artembilan commented on src/main/java/org/springframework/integration/aws/support/S3FileInfo.java:
```diff
@@ -0,0 +1,69 @@
+	private S3ObjectSummary s3ObjectSummary;
+
+	public S3FileInfo(S3ObjectSummary s3ObjectSummary) {
+		this.s3ObjectSummary = s3ObjectSummary;
```
`Assert.notNull(s3ObjectSummary, "s3ObjectSummary must not be null");`

artembilan commented on src/main/java/org/springframework/integration/aws/support/S3FileInfo.java:
```diff
@@ -0,0 +1,69 @@
+ */
+public class S3FileInfo extends AbstractFileInfo<S3ObjectSummary> {
+
+	private S3ObjectSummary s3ObjectSummary;
```
`final`

artembilan commented on src/main/java/org/springframework/integration/aws/support/S3FileInfo.java:
```diff
@@ -0,0 +1,69 @@
+
+	@Override
+	public boolean isDirectory() {
+		return (s3ObjectSummary != null ? s3ObjectSummary.getKey().endsWith("/") : false);
```
Can be simplified.
And I'm not sure that `/` is panacea to determine the dir...

The S3 has one dir - `bucket`. Everything rest is keys and when we scan the `bucket` for files we receive the entire path, even if it is artificial sub-dir:

```
bucket/foo.txt
bucket/bar/foo.txt
bucket/bar/baz/foo.txt
```
will be listed only once, without any recursion. So, we will never receive something like `bucket/bar/`

Therefore this method should always return `false`.

I won't mind if I miss something.
Thanks

artembilan commented on src/main/java/org/springframework/integration/aws/support/S3FileInfo.java:
```diff
@@ -0,0 +1,69 @@
+
+	@Override
+	public String getFilename() {
+		return (s3ObjectSummary != null ? s3ObjectSummary.getKey() : null);
```
After `Assert.notNull()` in ctor all this expressions can be simplfied.

artembilan commented on src/main/java/org/springframework/integration/aws/support/S3FileInfo.java:
```diff
@@ -0,0 +1,69 @@
+
+	@Override
+	public String getPermissions() {
+		return "-rw-r--r--";
```
Not sure in this hard code.
Maybe `AmazonS3.getObjectAcl(String bucketName, String key)` could be useful for us ?

tzolov replied:
The hard coding was fast and dirty hack. Need to test against real S# if the getObjectAcl will work before applying it.

tzolov replied:
I am afraid that even if getObjectAcl can do the trick it will be extremely expensive! It will require an additional REST call for each object in the bucket list. And this will be repeated at each pull.  
At the same time the benefit of having the exact permissions isn't that obvious.

artembilan replied:
Yep! That was my thought too.
So, let's make this method like:
```
@Override
public String getPermissions() {
	throw new UnsupportedOperationException("Use [AmazonS3.getObjectAcl()] to obtain permissions."); 
}
```
Plus we have to override `toString()` from super:
```
@Override
	public String toString() {
		return "FileInfo [isDirectory=" + isDirectory() + ", isLink=" + isLink()
				+ ", Size=" + getSize() + ", ModifiedTime="
				+ new Date(getModified()) + ", Filename=" + getFilename()
				+ ", RemoteDirectory=" + getRemoteDirectory() + "]";
	}
```
WDYT?

tzolov replied:
I like it! Wasn't sure if the UOE is acceptable response that doesn't break the rest of the workflow.

artembilan replied:
Well, look, we do this here in S3:
```
AbstractFileInfo<F> file = poll();
if (file != null) {
	String remotePath = remotePath(file);
	Session<?> session = this.remoteFileTemplate.getSession();
	try {
        	return getMessageBuilderFactory().withPayload(session.readRaw(remotePath))
```
So, we really don't care about permissions.

OTOH, the `AbstractRemoteFileOutboundGateway` returns `List<AbstractFileInfo<F>>` for `LS` command as a `payload`.
So, that is a target application responsibility to be ready for such a behavior on method during payload processing in its code. 
Like we can't remove from the `SingletonList` and so on.
But I guess we have to mention that in this method JavaDocs.

artembilan commented on src/main/resources/org/springframework/integration/aws/config/spring-integration-aws-1.0.xsd:
```diff
@@ -433,6 +433,144 @@
+							<xsd:documentation><![CDATA[
+								Specify a Comparator to be used when ordering Files. If none is provided, the
+								order in which files are processed is the order they are received from the
+								SFTP server. The generic type of the Comparator must be 'S3FileInfo'.
```
Typo: s/SFTP server/Amazon S3

artembilan commented:
That's all for this round.
Really cool work :smile: 

Thank you very much for such a valuable contribution!

tzolov commented:
@artembilan , i've addressed above suggestions except the `getPermissions`.  
Tomorrow will run manual tests to verify that `getObjectAcl` would do the job.

artembilan commented:
Excellent!

Please, don't squash commits.
We do that on merge anyway, but having several commits in the PR helps for review and for history.

artembilan commented on README.md:
```diff
@@ -115,6 +115,78 @@ An XML variant may look like:
 
+###Streaming Inbound Channel Adapter
+
+This adapter produces message with payloads of type `InputStream`, allowing S3 objects to be fetched without writing to the local file system. Since the session remains open, the consuming application is responsible for closing the session when the file has been consumed. 
```
It is better to keep each sentence on its own line: http://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

tzolov commented:
fine, i've misunderstood the commit instructions:  https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc#squash-commits

artembilan commented:
Ah! Sorry for confuse. That means for the first commit for initiating Pull Request.
Will fix that. 
Thank you!

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/S3InboundStreamingMessageSource.java:
```diff
@@ -0,0 +1,61 @@
+		super(template, null);
+	}
+
+	public S3InboundStreamingMessageSource(RemoteFileTemplate<S3ObjectSummary> template, Comparator<AbstractFileInfo<S3ObjectSummary>> comparator) {
```
No more then 120 symbols per code line.
Pay attention that it is a bit inconvenient to review it here on GH.

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/S3InboundStreamingMessageSource.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * A {@link AbstractRemoteFileStreamingMessageSource} implementation for the Amazon S3.
+ *
+ * @author Christian Tzolov
```
`@since 1.1` for all new public API.

tzolov replied:
Does `S3FileInfo` also qualify as a public API?

artembilan replied:
Yes, fully similar to `(S)FtpFileInfo`.
I don't see reason to hide our classes from end-users.
They might be useful somehow.
And that is end-user responsibility to treat then on his own.

And yes: new test classes must be be marked with `@since` as well.

artembilan commented on src/main/java/org/springframework/integration/aws/support/S3FileInfo.java:
```diff
@@ -0,0 +1,72 @@
+
+	@Override
+	public long getSize() {
+		return s3ObjectSummary.getSize();
```
Hm. Looks like our Checkstyle isn't full.
`this.` is required for class properties.

artembilan replied:
So, yes, I found that Checkstyle has been misconfigured.
Need fix to this in the `build.gradle`:
```
checkstyle {
	configFile = file("${rootDir}/src/checkstyle/checkstyle.xml")
	toolVersion = "6.16.1"
}
```
But, eh, not your task too :smile: 

Although it looks like we are pretty close to the finish :wink:

tzolov replied:
I've added already the missing `this`,  shall I add the checkstyle rule in the same commit too?

artembilan replied:
That would be great!
And `gradlew clean check` should confirm that we are good to go ahead.
Thank you!

artembilan commented on src/main/resources/org/springframework/integration/aws/config/spring-integration-aws-1.1.xsd:
```diff
@@ -433,6 +433,144 @@
+		</xsd:complexType>
+	</xsd:element>
+
+	<xsd:complexType name="base-inbound-adapter-type">
```
I guess we can reuse this new type from the `s3-inbound-channel-adapter`.
Nice!

But, right, it can be addressed in separate commit/PR.

tzolov replied:
Yes, I borrowed the type definitions from the sftp streaming project. Didn't want to modify the existing type `s3-inboud-channel-adapter`, but I can give it a try :)

artembilan replied:
:+1: 
But I won't insist.
If we agree with `UOE` for permissions, I can do the rest of polishing during merge.
For me the PR looks ready for merge.

artembilan commented on src/test/java/org/springframework/integration/aws/config/xml/S3InboundStreamingChannelAdapterParserTests.java:
```diff
@@ -0,0 +1,103 @@
+	private SourcePollingChannelAdapter s3Inbound;
+
+	@Autowired
+	private Comparator<?> comparator;
```
This property is out of use.
I guess just missed assertion.

artembilan commented on src/test/java/org/springframework/integration/aws/config/xml/S3InboundStreamingChannelAdapterParserTests.java:
```diff
@@ -0,0 +1,103 @@
+		S3PersistentAcceptOnceFileListFilter filter = TestUtils.getPropertyValue(source, "filter",
+				S3PersistentAcceptOnceFileListFilter.class);
+		assertThat(filter).isNotNull();
+		assertThat(filter).isEqualTo(acceptOnceFilter);
```
`isSameAs()`  would be better because we are asserting exactly injection.

artembilan replied:
And, therefore, the previous `isNotNull()` redundant.

artembilan commented on src/test/java/org/springframework/integration/aws/inbound/S3InboundStreamingChannelAdapterTests.java:
```diff
@@ -0,0 +1,194 @@
+
+			S3SessionFactory s3SessionFactory = new S3SessionFactory(amazonS3);
+
+			S3RemoteFileTemplate s3FileTempalte = new S3RemoteFileTemplate(s3SessionFactory);
```
Typo: s3FileTemp**la**te
Sorry, but my IDEA doesn't like my English so I switched on Spelling plugin
:smile:

artembilan commented:
That's all!
Waiting for your polishing and will be happy to merge

tzolov commented:
Ok, i've addressed all remarks but couldn't make the `s3-inbound-channel-adapter` extend the same `base-inbound-adapter-type`.

artembilan commented:
Checkstyle problem is like this: https://github.com/checkstyle/checkstyle/issues/3418
So, we don't have choice unless still sit with version `6.16.1`.

Merged as https://github.com/spring-projects/spring-integration-aws/commit/da77e4b89e2ca15f6c22b374dd654c2e4cf5da16, BTW.

That was perfect co-operation.

Looking forward for more!

tzolov commented:
Glad I had cache to work on this :) Thanks for the guidelines and all the tips! 
Next, i guess is to turn the streaming s3 source into `spring-cloud-stream-app-starters`. Is there an open ticket in the starters project or shall i create new one?
